daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala

```diff
@@ -926,14 +920,19 @@ object Ast {
   // Modules and packages
   //
 
-  case class GenModule[E] private[Ast] (
+  final case class GenModule[E](
       name: ModuleName,
       definitions: Map[DottedName, GenDefinition[E]],
       templates: Map[DottedName, GenTemplate[E]],
       exceptions: Map[DottedName, GenDefException[E]],
       interfaces: Map[DottedName, DefInterface],
       featureFlags: FeatureFlags,
-  ) extends NoCopy
+  ) {
+    templates.keysIterator.foreach(name =>
+      if (exceptions.keySet.contains(name))
+        throw PackageError(s"Collision between exception and template name ${name.toString}")
+    )
+  }
```

Comment:
Once putting the invariant check inside `GenModule`, we do not need the constructor to be private